motorway/contrib/kafka/ramps.py

```diff
@@ -0,0 +1,136 @@
+import logging
+import threading
+import time
+from collections import defaultdict
+from queue import Queue
+
+from confluent_kafka import Consumer, TopicPartition
+from confluent_kafka import Message as KafkaMessage
+
+from motorway.contrib.kafka.mixins import KafkaMixin
+from motorway.messages import Message
+from motorway.ramp import Ramp
+
+logger = logging.getLogger(__name__)
+
+class KafkaRamp(Ramp, KafkaMixin):
+    topic_name = None
+    AUTO_OFFSET_RESET = 'latest'
+    MAX_UNCOMPLETED_ITEMS = 3000
+    GET_RECORDS_LIMIT = 1000
+    THROTTLE_SECONDS = 5
+
+    def __init__(self, consumer_thread_enabled=True, consume_iterations=None, **kwargs):
+        super().__init__(**kwargs)
+        assert self.topic_name, "Please define the attribute `topic_name` for your KafkaRamp"
+
+        self.insertion_queue = Queue()
+        self.uncompleted_ids = defaultdict(set)
+
+        self.consumer = Consumer({
+            **self.connection_parameters(),
+        })
+        if consumer_thread_enabled:
+            self.start_consumer_thread()
+        else:
+            self.consume(iterations=consume_iterations)
+            logger.info('Skipping consumer thread for topic %s', self.topic_name)
+
+    def start_consumer_thread(self):
+        thread = threading.Thread(
+            target=self.consume,
+            name=f"{self.__class__.__name__}-consume-{self.topic_name}",
+        )
+        logger.info("Starting Kafka consumer thread for topic %s", self.topic_name)
+        thread.start()
+
+    def _too_many_uncompleted_items(self):
+        """
+        Pause consumption if we have too many uncompleted items
+        because we use kafka built-in balancing of consumer group, we need to pause consumption
+        from all assigned partitions even if only one of them has too many uncompleted items.
+        """
+        return any(
+                len(self.uncompleted_ids[partition]) > self.MAX_UNCOMPLETED_ITEMS
+                for partition in self.uncompleted_ids.keys()
+        )
+
+    def _throttle(self):
+        logger.warning("Too many uncompleted items, pausing consumption for %d seconds", self.THROTTLE_SECONDS)
+        time.sleep(self.THROTTLE_SECONDS)
+
+    def consume(self, iterations=None):
```

Comment:
Let's wrap this with decorator to catch exception when consumer is killed and re-run it.